DM-14406: Initial documentation set

jonathansick opened the pull request:
This is the initial documentation set for the LSP Notebook Aspect.

**Preview: https://nb.lsst.io/v/DM-14406/**

Please provide suggestions on this PR! Typos, fact checks, and issues with usability of the docs are all appreciated.

Keep in mind that we're purposefully **not** documenting the JupyterLab UI in these LSST docs. At this stage, we're linking to the JupyterLab docs for that. This strategy helps prevent these docs from going out of date because of upstream UI changes.

jonathansick committed:
Add conf.py to configure the Sphinx build

jonathansick committed:
Add requirements.txt

Soon I'll want to pin these dependencies a bit better.

jonathansick committed:
Add sphinx Makefile

This Makefile is highly stripped-down from the typical Sphinx Makefile
because I only want to support the html builder. Otherwise people will
request fixes to unsupported EPUB and latex builds :)

jonathansick committed:
Add placeholder _static dir

This directory is expected by Sphinx.

jonathansick committed:
Add gitignore for _build

jonathansick committed:
Add index.rst

jonathansick committed:
Deploy to nb.lsst.io with ltd-conveyor

jonathansick committed:
Remove colon from LSP Notebook Aspect name

Without the colon the full name will work better in the text as well.

jonathansick committed:
Add Logging in page

This topic describes how to log into the Notebook, and includes Cisco
AnyConnect info.

jonathansick committed:
Add Logging out page

jonathansick committed:
Add Getting help topic

jonathansick committed:
Add topics on notebook-demo and other repos

- The notebook-demo repo is special because it's preinstalled, so I'm
  featuring it in the docs.
- The "More notebook repositories" page is meant to be a helpful index
  of interesting LSST-related notebook repositories on GitHub.

jonathansick committed:
Add topics on pipelines in notebooks and terminal

- Covers sourcing loadLSST in the terminal and setup lsst_distrib
- Covers ~/notebooks/.user_setups for notebooks

jonathansick committed:
Add Filesystem directories topic

This page describes what directories are available from the notebook
aspect.

jonathansick committed:
Add topic on the Python environment

Covers installed packages and how to install additional packages.

jonathansick committed:
Add topic on installed software

The listing is exhaustive. Its more descriptive to suggest what tools
are available. I think this slightly imprecise approach is more
maintainable as the Docker image gets updated.

jonathansick committed:
Add reference topic on JupyterLab extensions

Users can't change these, but they might find it useful to know that
they're installed.

jonathansick committed:
Add shell configuration topic

This is mostly a reminder that .bashrc/.bash_profile exist and can be
customized. In the future the page might be expanded with specific
configuration tips.

jonathansick committed:
Add topic on Git configuration

This topic is intended to ensure that users are configuring git
successfully to collaborate on notebooks through GitHub.

- Basic set up
- Credential helper
- Git LFS basics

jonathansick committed:
Add a notebook development tutorial

This is based on Simon Krughoff's https://sqr-021.lsst.io technote. I've
changed it up a bit to encourage using real text editors rather than
"cheating" with sed because coming to grips with editing text on
JupyterLab is an essential part of development.

jonathansick committed:
Add screenshots of the demo notebooks

I couldn't get a screenshot for the Firefly notebook because it isn't
currently working.

Process:

- macOS screenshot from Safari.
- Compress the screenshot in Photoshop, but using full resolution (rely
  on compression to reduce image size).

jonathansick committed:
Add dome image to homepage

jonathansick committed:
Use lsst-sphinx-bootstrap-theme 0.2.0rc1

jonathansick committed:
Configure linkcheck to ignore lsst-lspdev URL

Can't rely on the endpoint being available because of the VPN
requirement.

SimonKrughoff commented:
Only one of two next steps at the bottom of the "Logging in" page is a link.

SimonKrughoff commented on getting-started/logging-out.rst:
```diff
@@ -0,0 +1,19 @@
+###########
+
+Once you’re ready to take a break, log out from the Notebook Aspect to release computational resources to others.
+Don’t worry, all your files — and even the layout of tabs — are saved. When you log in later, you can pick right up.
```
Note that only notebook tabs are saved from session to session.  Terminal tabs are not saved from session to session.

jonathansick committed:
Add link to demos in next steps after log in

stargaser commented on notebooks/repositories.rst:
```diff
@@ -0,0 +1,51 @@
+.. _`lsst-sqre/notebook-demo`: https://github.com/lsst-sqre/notebook-demo
+.. _`lsst-com/notebooks`: https://github.com/lsst-com/notebooks
+.. _`LSSTScienceCollaborations/DMStackClub`: https://github.com/LSSTScienceCollaborations/DMStackClub
+.. _`lsst-dm-tutorial/lsst2017`: https://github.com/lsst-dm-tutorial/lsst2017
```
This one was moved to https://github.com/lsst-dm/tutorial-lsst2017 ... perhaps more importantly, the stack no longer works on the tarfile that seeds this tutorial.

SimonKrughoff commented on notebooks/repositories.rst:
```diff
@@ -0,0 +1,51 @@
+.. _`LSSTScienceCollaborations/DMStackClub`: https://github.com/LSSTScienceCollaborations/DMStackClub
+.. _`lsst-dm-tutorial/lsst2017`: https://github.com/lsst-dm-tutorial/lsst2017
+
+Suggest a repository
```
FYI, I think DESC will be having notebooks shortly.

stargaser commented on science-pipelines/science-pipelines-in-notebooks.rst:
```diff
@@ -0,0 +1,85 @@
+When the LSST kernel starts up, it activates the LSST environment and sets up these packages:
+
+-  ``lsst_distrib`` (the top-level EUPS package of the `LSST Science Pipelines`_)
+-  ``display_firefly`` (an additional package that enables you to use Firefly from the display framework).
```
N.B. `display_firefly` is included in `lsst_distrib`.

SimonKrughoff commented on science-pipelines/science-pipelines-in-notebooks.rst:
```diff
@@ -0,0 +1,82 @@
+Switch an opened notebook to the LSST kernel
+============================================
+
+You can open existing notebooks by double clicking on their names in the JupyterLab file browser.
```
FYI, I've run into this several times.  If you double click the name, it will try to rename the file.  The foolproof thing is to double click icon to the left.

SimonKrughoff commented on science-pipelines/development-tutorial.rst:
```diff
@@ -0,0 +1,173 @@
+.. code-block:: python
+
+   import lsst.pipe.tasks
+   print(lsst.pipe.tasks.__file__)
```
I find it useful to source the `.user_setups` in the terminal just to make sure a bash error doesn't get silently ignored.

SimonKrughoff commented on environment/directories.rst:
```diff
@@ -0,0 +1,84 @@
+
+Your home directory is just that.
+It’s backed by the same storage as your home directory on `lsst-dev`_ in the LSST Data Facility.
+In fact, your :file:`$HOME` directory in the Notebook Aspect *is* the :file:`$HOME/jhome/` directory on `lsst-dev <https://developer.lsst.io/services/lsst-dev.html>`_. 
```
`$HOME/jhome/` is actually a symlink to `/lsst/jhome/$USER`.  We should double check with Michelle that these have the same policies as `$HOME`. (unless you already have)

SimonKrughoff reviewed (approved):
Looks great.

stargaser commented on getting-started/logging-in.rst:
```diff
@@ -0,0 +1,76 @@
+
+Next, click **Spawn** and you’ll be launched into the Notebook Aspect.
+
+Next steps
```
I am a bit confused by the Next Steps here. The Next menu at the top goes to logging out. Learning more about Jupyterlab takes me out of these docs (maybe want to open in a new tab?). Explore demo notebooks is suggested...should I review logging out?

For a long page like this, I also wish that the Previous and Next links at the top were repeated at the bottom.

jonathansick replied:
Thanks, I hadn't thought about that UX. I'll see what I can do about the nav from the theme side.

In terms of site structure, it sounds like I should be moving the pages on `Demo notebooks` and `More notebook repositories` up to the Getting started part, right after Logging in. That would be a better initial experience than going straight to covering logging out and getting help. Probably no need for a dedicated section in the docs on sample notebooks.

stargaser commented on getting-started/logging-in.rst:
```diff
@@ -0,0 +1,76 @@
+
+Open `lsst-lspdev.ncsa.illinois.edu/nb <https://lsst-lspdev.ncsa.illinois.edu/nb>`__ in your web browser. Click the **Sign in with CILogon** button, then enter your NCSA Kerberos credentials on the NCSA CILogon page.
+
+Once you click **Log in**, you’ll be redirected to the Notebook Aspect’s Spawner page.
```
I have an extra authentication step, Duo, to have special access to the PDAC. Maybe better to say "After authentication is completed" instead of "Once you click Log In", to cover all scenarios.

stargaser commented on getting-started/logging-out.rst:
```diff
@@ -0,0 +1,19 @@
+Once you’re ready to take a break, log out from the Notebook Aspect to release computational resources to others.
+Don’t worry, all your files — and even the layout of tabs — are saved. When you log in later, you can pick right up.
+
+The best way to log out is by clicking on **File** → **Save All, Exit, and Log Out**.
```
Very happy to see this advice! I thought I was supposed to do Kernel -> Log Out and I kept forgetting to do that.

stargaser commented on science-pipelines/science-pipelines-in-notebooks.rst:
```diff
@@ -0,0 +1,82 @@
+
+When the LSST kernel starts up, it activates the LSST environment and sets up ``lsst_distrib`` (the top-level EUPS package of the `LSST Science Pipelines`_).
+
+You can customize what packages are set up by creating and editing a :file:`~/notebooks/.user_setups` file.
```
I just confirmed that environment variables can be set here too. They are picked up after restarting the kernel. Excellent!

stargaser replied:
Might be good to add that restarting the kernel will pick up the changes.

stargaser commented on science-pipelines/science-pipelines-in-notebooks.rst:
```diff
@@ -0,0 +1,82 @@
+
+The :file:`~/notebooks/.user_setups` file is sourced by bash.
+This means you can include any valid bash syntax.
+For example, you can define environment variables that are accessible from notebooks.
```
Oh, here is where environment variables are mentioned...maybe these should be more prominent, especially now that we are defaulting Firefly servers using them.

stargaser commented on environment/shell-configuration.rst:
```diff
@@ -0,0 +1,13 @@
+
+.. note::
+
+   Environment variables that you set in :file:`~/.bashrc` or :file:`~/.bashrc_profile` are not available from notebooks.
```
Good to know! I wonder if this can be repeated or linked somehow to the section on notebooks/.user_setups.

stargaser reviewed (approved):
Looks terrific to me! A few minor in-line comments, mostly about environment variables...it all became clear as I went through all the sections.

jonathansick committed:
Add tip to source user_setups to check it

This is currently the only way to alert a user that there's a syntax
error in their ~/notebooks/.user_setups file.

jonathansick committed:
Add changing user_setups in running kernels

jonathansick committed:
Emphasize use of user_setups for env vars

- rename title of section to explicitly suggest export environment
  variables.

- include an example export for emphasis.

- Better linking between this section and the section about bash
  configuration.

jonathansick committed:
Improve science pipelines-in-notebooks into sec

Mentioning user_setups in the intro section will hopefully add some
visibility to it given that its the last section on the page and is
below the fold.

jonathansick committed:
Mention deployment URL on homepage

Since this documentation is specific to lsst-lspdev, but there are many
deployments that differ in subtle ways, it's good to mention the
deployment on the homepage.

Soon we'll want to dynamically customize the documentation for different
deployments.

jonathansick committed:
Consolidate notebook section to getting-started

This moves the topics on demo notebooks and links to other notebook
repos to the Getting started section:

- Improve initial experience because you're guided towards looking at
  existing notebooks to learn, instead of going straight to the "Logging
  out" page.

- I'm uncertain about whether a dedicated section on notebook repos
  really has any useful purpose.

jonathansick committed:
Improve page title of notebook repos topic

jonathansick committed:
Improve link in next steps section

- emphasize the JupyterLab docs are external.
- make the link target sizes more uniform.

drphilmarshall commented:
Great stuff - thanks for making this, @jonathansick ! It looks really good to me. The only thing I can think of that it might be good to add is a note on how to contribute improvements to the notebooks in `notebook-demo`. You say that commits to `prod` should be avoided, but presumably if we checkout a dev branch you are open to pull requests... If you do add a note on this, you might remind people to clear the outputs first.

jonathansick commented:
@drphilmarshall Good point. I'm not 100% sure about the contribution workflow for notebook-demo. @SimonKrughoff, do you want to say anything about that at this stage?

jonathansick committed:
Use shorter title for cookie crumb nav bar

jonathansick committed:
Use lsst-sphinx-bootstrap-theme==0.2.0rc4

jonathansick committed:
Tweak intro sentence

jonathansick committed:
Add topic with links to project docs and repos

Especially early-on, many types of users will want to learn more about
the notebook aspect. These are the essential project docs.

jonathansick committed:
Add documentation on contributing to the docs

- How to make issues
- How to make PRs
- How to build the docs
- Style guide

This also updates the relevant .github pages.

jonathansick committed:
Add a page with the documentation license

jonathansick commented:
I've added a new [project info](https://nb.lsst.io/v/DM-14406/index.html#project-info) section to the bottom of the docs:

- [Implementation documentation](https://nb.lsst.io/v/DM-14406/project/implementation.html) — find associated LDMs and technotes.
- [Contributing to the docs](https://nb.lsst.io/v/DM-14406/project/contributing.html) — sets some contribution guidelines and pointers.
- [Documentation license](https://nb.lsst.io/v/DM-14406/project/license.html) — describes the CC-BY license.

jonathansick committed:
Note 100GB quota for home directories

See https://lsstc.slack.com/archives/C2K6D7R61/p1526942695000274

drphilmarshall commented:
Looks good, @jonathansick - perhaps its enough then for the `notebook-demo` README to have a "Contributing" section? I just proposed one in https://github.com/lsst-sqre/notebook-demo/pull/6

jonathansick committed:
Use theme 0.2.0rc5

jonathansick committed:
Use lsst-sphinx-bootstrap-theme 0.2 releases

jonathansick committed:
Document VPN issues and workaround

I've noticed in Slack people having issues when presented with the VPN
installation. This text is intended to reassure users that the failures
are expected and to persist through them.

jonathansick committed:
Tweak paragraph on onboarding/getting accounts

- supervisor -> sponsor
- fix typo.
- de-emphasize DM onboarding
